accepted/2022/capability-analyzer/capability-analyzer.md

```diff
@@ -0,0 +1,205 @@
+# Capability APIs
+
+**PM** [Immo Landwerth](https://github.com/terrajobst)
+
+*This is a revamped version of a [previous draft](https://github.com/dotnet/designs/pull/111).*
+
+In .NET 6, we introduced the [Platform Compatibility Analyzer][platform-compat].
+It allows us to mark platform specific APIs. This works well for APIs that are
+intrinsically platform specific (e.g. the Windows registry) or are wrappers
+around OS specific APIs that are only implemented for a specific set of
+platforms. The idea is that consumers will either mark themselves as platform
+specific or dynamically check by calling, for example,
+`OperatingSystem.Is<Platform>()`.
+
+For cross-cutting concerns, such as code generation or threading, this pattern
+doesn't work super well because the list of platforms can change over time or
+because the platform concept isn't sufficient, when, for example, the runtime
+behavior depends on orthogonal choices such as when the code is compiled
+ahead-of-time.
+
+## Scenarios and User Experience
+
+### Thread Creation
+
+Maddie is building a Blazor WebAssembly application that she is porting from a
+legacy Windows desktop application. Part of the application is uploading
+relatively large data files to a cloud backend. To avoid disappointment down the
+road, the UI performs some upfront validation that parses the input files. To
+prevent the UI from freezing up, someone wrote some code that offloaded that
+logic to a worker thread. While that is not ideal, that worked well for the last
+decade in the .NET Framework app. She copy and pastes a good chunk of code that
+includes the following snippet:
+
+```C#
+void StartValidation()
+{
+    InputValidationConfiguration args = GetConfiguration();
+    SynchronizationContext syncContext = SynchronizationContext.Current;
+
+    Thread thread = new Thread(() =>
+    {
+        var result = Validator.Validate(args);
+        syncContext.Post(_ => ValidationComplete(result));
+    });
+    
+    thread.Start();
+}
+```
+
+She immediately gets a diagnostic:
+
+    CAXXXX: The API 'Thread.Start()' requires the capability 'ThreadCreation'. Either mark the consumer with `[RequiresThreadCreation]` or check by calling 'Thread.IsCreationSupported'.
```

Comment:
Are there other APIs except `Thread.Start` that would be marked as requiring the thread create capability?

Reply:
I see that `Task.WaitAll` is marked as supported on browser, but `Task.Wait` is not. Why is that?

I would expect `Task.WaitAll` to be just a more efficient version of `foreach (Task t in tasks) t.Wait()`.

Are both `Task.Wait` and `Task.WaitAll` going to be marked with this capability?

Reply:
I believe the idea behind is that if you are in a single-threaded setup Task.WaitAll will never return as you wait on more than 1 concurrent thing (yeah there are some exceptions). Therefore if you are calling it you are doing something wrong.

/cc @lewing

Reply:
> I believe the idea behind is that if you are in a single-threaded setup Task.WaitAll will never return as you wait on more than 1 concurrent thing (yeah there are some exceptions).

That should not be the case.  As Jan suggested, behaviorally there shouldn't be a difference between using Task.WaitAll on multiple things and using Task.Wait on each of those multiple things.

Further, both WaitAll and Wait can "inline" the thing they're waiting on if it's backed by a delegate to be executed (e.g. Task.Run(action)), so if you are in a single-threaded environment and you're waiting on such tasks, they should both "just work".  If they don't in such cases, that suggests we're missing handling of this case in browser.

Reply:
I'm surprised WaitAll isn't marked there are definitely certain circumstances where it can block.  I suspect it got overlooked.

> 
> Further, both WaitAll and Wait can "inline" the thing they're waiting on if it's backed by a delegate to be executed (e.g. Task.Run(action)), so if you are in a single-threaded environment and you're waiting on such tasks, they should both "just work". If they don't in such cases, that suggests we're missing handling of this case in browser.

In the browser we have tasks (e.g. any JS interop involving a Promise) that cannot complete without yielding execution back to the browser main loop, so they won't "just work".   Wait and WaitAll _can_ work under many circumstances but  in practice they can and do break in confusing ways.

Reply:
> so they won't "just work"

Right, that's the category I mentioned where they're not backed by delegates.

Both WaitAll and Wait would have the same issue there; WaitAll isn't special in that regard.